### ISSUE TITLE
[Bug] Numerically Unstable MultiLabelSoftMarginLoss

### ISSUE THREAD

**Issue by Jiaming-Liu** [pytorch/pytorch#1516]
Please see this test: the gradient goes wrong in some test cases (obtained 0.0000 vs expected 0.1000).
https://github.com/Jiaming-Liu/pytorch-misc/blob/master/test_binary_cross_entropy.ipynb

cc @SsnL

**Comment by Jiaming-Liu** [comment]
You might also see the same exp. on tensorflow:
https://github.com/Jiaming-Liu/pytorch-misc/blob/master/test_tf_binary_cross_entropy.ipynb

BTW, it seems that torch.nn.MultiLabelSoftMarginLoss scales down the gradient by the number of classes by default. Is there any particular reason for it?

**Comment by Jiaming-Liu** [comment]
https://github.com/Jiaming-Liu/pytorch-misc/blob/master/sigmoid_with_binary_cross_entropy.py

This can be a temporary fix for this issue.

**Comment by oleg-trott** [comment]
Looks related to https://github.com/torch/nn/issues/949

**Comment by Jiaming-Liu** [comment]
@apaszke What is the current status of this issue? In torch, a shortcut is token to avoid this numerical issue.

**Comment by Coderx7** [comment]
Is this fixed?

**Comment by mruberry** [comment]
This appears to have been fixed. Following the provided script, a recent comparison is:

(actual)
```
(0, tensor(-0.0475))
(1, tensor(-0.0525))
(2, tensor(0.1000))
(3, tensor(0.1000))
(4, tensor(0.1000))
(5, tensor(0.1000))
(6, tensor(0.0500))
(7, tensor(3.6955e-07))
(8, tensor(0.))
(9, tensor(0.))
```

vs.

(expected)
```
(0, tensor(-0.0475))
(1, tensor(-0.0525))
(2, tensor(0.1000))
(3, tensor(0.1000))
(4, tensor(0.1000))
(5, tensor(0.1000))
(6, tensor(0.0500))
(7, tensor(3.7266e-07))
(8, tensor(1.3888e-12))
(9, tensor(1.9287e-23))
```
